packages/control/algorithm/surplus_controlled.py

```diff
@@ -3,18 +3,20 @@
 
 from control import data
 from control.algorithm import common
+from control.algorithm.chargemodes import CHARGEMODES
 from control.loadmanagement import LimitingValue, Loadmanagement
 from control.counter import Counter
 from control.chargepoint.chargepoint import Chargepoint
-from control.algorithm.filter_chargepoints import (get_chargepoints_by_mode, get_chargepoints_by_mode_and_counter,
+from control.algorithm.filter_chargepoints import (get_chargepoints_by_chargemodes,
+                                                   get_chargepoints_by_mode_and_counter,
                                                    get_preferenced_chargepoint_charging)
 from control.chargepoint.chargepoint_state import ChargepointState, CHARGING_STATES
 from modules.common.utils.component_parser import get_component_name_by_id
 
 log = logging.getLogger(__name__)
 
-CONSIDERED_CHARGE_MODES = common.CHARGEMODES[0:2] + common.CHARGEMODES[6:12]
-CONSIDERED_CHARGE_MODES_PV = common.CHARGEMODES[8:12]
+CONSIDERED_CHARGE_MODES = CHARGEMODES[0:2] + CHARGEMODES[6:12]
+CONSIDERED_CHARGE_MODES_PV = CHARGEMODES[8:12]
```

Comment:
Ich bin kein Python-Experte, aber den Zugriff mit den Indizes finde ich sehr fehleranfällig. Ich vermute, dass es da eine bessere Variante gibt.

So als erste Idee könnte man die reine Liste `CHARGEMODES` durch eine Klasse ersetzen. Mit einzelnen Methoden kann dann jeweils eine definierte Teilmenge zurückgegeben werden. Dann wären zumindest die Indizes in der selben Datei/Klasse wie die ursprüngliche Listendefinition. Code ausserhalb der Datei/Klasse braucht dann bei eventuellen Änderungen nicht aufwendig gesucht und angepasst werden.

Reply:
Ich habe die Definition der Konstanten in die Datei von `CHARGEMODES` verschoben. So ist alles zusammen. 

Einen Vorteil, `CHARGEMODES` in eine Klasse umzuwandeln und die Konstanten in Methoden, sehe ich nicht. `CHARGEMODES` ist ein Tupel, dh die Reihenfolge ist unveränderlich. Darüber wird die Priorität der Lademodi definiert. Je kleiner der Index, desto höher die Priorität. Auch in einer Klasse müsste das Attribut ein Tupel bleiben. Methoden statt Konstanten zu verwenden, bringt Overhead, da die Variablen (=keine) und der Rückgabewert (=Konstante) schon vorher festgelegt sind.